language/vm/src/abi.rs

```diff
@@ -0,0 +1,399 @@
+/// want a change from `Account { bal: u64, seq: u64 }` to `Account { seq: u64, bal: u64 }` to be
+/// marked as incompatible. Not safe to compare without an enclosing `StructABI`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FieldABI {
+    pub name: Identifier,
+    pub type_: TypeABI,
+}
+
+/// ABI for a kind constraint on a generic type parameter
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum KindABI {
+    All,
+    Resource,
+    Copyable,
+}
+
+/// ABI for a struct declaration. Not safe to compare without an associated `ModuleId` or `ModuleABI`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StructABI {
+    pub name: Identifier,
+    pub kind: KindABI,
+    pub type_parameters: Vec<KindABI>,
+    pub fields: Vec<FieldABI>,
+}
+
+/// ABI for a function declaration. Not safe to compare without an associated `ModuleId` or `ModuleABI`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FunctionABI {
+    pub name: Identifier,
+    pub type_parameters: Vec<KindABI>,
+    pub formals: Vec<TypeABI>,
+    pub ret: Vec<TypeABI>,
+}
+
+/// ABI for a `CompiledModule`: its address, name, struct declarations, and public function
+/// declarations.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct ModuleABI {
+    pub address: AccountAddress,
+    pub name: Identifier,
+    pub structs: Vec<StructABI>,
+    pub public_functions: Vec<FunctionABI>,
+}
+
+/// The result of an ABI compatibility check. Here is what the different combinations mean:
+/// `{ struct: true, struct_layout: true }`: fully backward compatible
+/// `{ struct_and_function_linking: true, struct_layout: false }`: Dependent modules that reference functions or types in this module may not link. However, fixing, recompiling, and redeploying all dependent modules will work--no data migration needed.
+/// `{ type_and_function_linking: true, struct_layout: false }`: Attempting to read structs published by this module will now fail at runtime. However, dependent modules will continue to link. Requires data migration, but no changes to dependent modules.
+/// `{ type_and_function_linking: false, struct_layout: false }`: Everything is broken. Need both a data migration and changes to dependent modules.
+pub struct Compatibility {
+    /// If false, dependent modules that reference functions or structs in this module may not link
+    pub struct_and_function_linking: bool,
+    /// If false, attempting to read structs previously published by this module will fail at runtime
+    pub struct_layout: bool,
+}
+
+impl Compatibility {
+    /// Return true if the two module ABIs compared in the compatiblity check are both linking and
+    /// layout compatible.
+    pub fn is_fully_compatible(&self) -> bool {
+        self.struct_and_function_linking && self.struct_layout
+    }
+}
+
+impl ModuleABI {
+    /// Return compatibility assessment for `new_module` relative to old module `self`
+    pub fn check_compatibility(&self, new_module: &ModuleABI) -> Compatibility {
```

Comment:
I assume the preconditions or assumptions on the verification status are similar to what's described in the doc? To me that's also a crucial part of this story.